src/openclsim/critical_path/dependencies_from_model.py

```diff
@@ -0,0 +1,243 @@
+        cross reference with recorded activities to get dependencies of relevant instances
+        requires self.object_list, self.activity_list and self.recorded_activities_df
+
+        Notes
+        ---------
+        this method is likely to provide too few dependencies and might raise warnings
+        for unsupported start event conditions, activities et cetera
+
+        this method contains approx 1000 (!) lines in original commit
+
+        Returns
+        -------
+        dependency_list : list of tuples
+            like [(A1, A2), (A1, A3), (A3, A4)] where A2 depends on A1 (A1 'causes' A2) et cetera
+        """
+
+        # get dependencies from DependencyGraph - so at model setup level (not time based yet)
+        my_graph = DependencyGraph(self.activity_list)
+        dependencies_act = my_graph.getListDependencies()
+
+        # also get dependencies from start events - HAS LIMITATIONS
+        dependencies_start = get_act_dependencies_start(self.activity_list)
+
+        # convert these dependencies to depencies of the activity 'instances',
+        # i.e. dependencies within the simulation/with time component
+        cp_dependencies = get_dependencies_time(
+            self.recorded_activities_df, dependencies_act + dependencies_start
+        )
+
+        # get cp dependencies based on resource utilisation vs capacity - HAS LIMITATIONS
+        cp_depencies_res_limitation = get_resource_capacity_dependencies(
+            self.recorded_activities_df, self.object_list
+        )
+
+        # get cp dependencies 'WAIT' - HAS LIMITATIONS
+        cp_dependencies_wait = get_wait_dependencies_cp(self.recorded_activities_df)
+
+        return list(
+            set(cp_dependencies + cp_depencies_res_limitation + cp_dependencies_wait)
+        )
+
+
+# Waiting related start events
+def get_wait_dependencies_cp(recorded_activities_df):
+    """
+    Get/assume dependencies between activities related to waiting. Waiting
+    activities must explicitly be marked in the CpLog object as ``WAITING``.
+
+    Parameters
+    -----------
+    recorded_activities_df : pd.DataFrame
+        attribute of base_cp.BaseCP
+
+    Returns
+    --------
+    list_wait_dependencies : list
+        list of tuples of dependencies as occured in simulation
+        (tuple of cp_activity_id) wrt WAITing
+
+    .. warning::
+        The definition of ``WAITING`` is up for discussion, as well as when and
+        whether waiting is part of the critical path in the first place.
+
+    .. warning::
+        This function is preliminary, and hence has its limitations. Mostly due
+        to critical information for dependencies in the way OpenCLSim now logs
+        its information. For proper dependency checks for extracting the
+        critical path, the logging module of OpenCLSim must be extended.
+
+    """
+    return []
+
+
+def get_resource_capacity_dependencies(recorded_activities_df, objects_list):
```

Comment:
We are making an object oriented module, no seperate functions! 

all the parameters of the functions are atributes of the the object. When you make the methods of the objects you dont have to pass the arguments. There is no reason the make these functions outside the object.

Reply:
ok - moved to class

Reply:
will be big one, but maybe for later discussion...